### ISSUE TITLE
chunked_map in seastar

### ISSUE THREAD

**Issue by travisdowns** [scylladb/seastar#2133]
Seastar cannot in general support large allocations after an application has been running for some time, due to its use of a custom allocator which allocate a fixed virtual memory region which never grows: in general the heap becomes fragmented and allocations some significant multiple of the page size are likely to fail.

Hence the existence of `chunked_fifo` which is a non-contiguous replacement for `std::vector` in some cases.

H/e, the same basic problem also arises with maps. In Redpanda, we will solve this problem using a "fragmented map" whose backing arrays are compose of chunks of some maximum size, not unlike `chunked_fifo` (except that we need random access, so there is a 2-level indexing scheme, not unlike `std::deque`).

Some of the large allocations can occur in seastar itself, e.g., if there are many metrics (link issue).

Given this, we propose that we upstream the fragmented map here instead of keeping in Redpanda. It would almost certainly be an existing 3rd party map implementation, with a compatible license, modified to be chunked. 

WDYT? 

Perhaps this problem already has a solution in Scylla: we'd be interested to hear about it if so.

**Comment by travisdowns** [comment]
cc @tchaikov any thoughts on this one?

**Comment by avikivity** [comment]
We don't have a fragmented unordered map (don't think we ever saw a need - to exceed 128k one would need 16k buckets, no?

In general I don't want Seastar to become a collections library.

**Comment by travisdowns** [comment]
> We don't have a fragmented unordered map (don't think we ever saw a need - to exceed 128k one would need 16k buckets, no?

Something like that: but this is possible for example (and in practice) if you have many metrics: 16K isn't an outlandish number of metrics. Another option would be to use something like `std::map`, which is inherently non-contiguous, for those use cases but performance of that implemetnation is ... not ideal.

**Comment by nyh** [comment]
> In general I don't want Seastar to become a collections library.

Yes, this is the current sentiment. Some historical context: We added chunked_fifo to Seastar almost eight years ago (commit e66ec715eae6e83d3c040429ace8566bdd9ae886) to use it **in** Seastar, namely the semaphore implementation. As you noticed, chunked_fifo is optimized to be just a queue, not a random-addressable vector, which is exactly what we needed for the semaphore.

Here's a core dump of my thoughts on this issue:

In the ScyllaDB project we did need a chunked **vector** type (with full random-access support), and added one (appropriately called "chunked_vector") but never tried to put it in Seastar. As you noted, you guys also need a chunked **map** type. One could argue that every serious Seastar user will likely need one or more of these chunked containers, so it does make sense to collect these chunked container implementations and put them into either Seastar itself or a separate library.

One reason to not want to put those container implementations in Seastar itself is that they are at all specific to Seastar - they don't use Seastar's futures, or any other Seastar function or header file. E.g., just like Boost have a set of "intrusive containers", it (or a new library) could have a set of "chunked containers".

Another, bigger problem, is how to choose which containers to include. Should we focus on chunked "versions" of existing C++ STL - e.g., a chunked version of std::map, a separate chunked version of std::unordered_map, etc.?  Or will we allow people to start adding all sort of alternative container implementations, each with different optimizations? There are many different container implementations in Boost, Google's libraries, and other places... Arguably, Seastar's existing chunked_fifo is already one of these "alternative" implementation - it is not 100% compatible with STL's std::dequeue because it's just a queue and does not implement O(1) random access. But as I said, we added because we needed to use it inside Seastar itself.

Another question to ponder about: Even if we only focus on "chunked versions of existing C++ STL containers", how stringent will we be (or can we be) about the complexity guarantees? For example, std::vector requires O(1) addressing. A chunked implementation cannot **really** have asymptotic (as N grows to infinity) O(1) performance. It can be log(N)/large_constant but not O(1). Alternatively, the "chunked container" can opt to guarantee <128KB allocations only up to a certain N. This is what Scylla's chunked_vector currently does: vectors up to 2GB have a maximum allocation size of 128KB, but larger vectors may need a contiguous allocation of more than 128KB. Is this fine? Probably, but as soon as we make this an "official" part of Seastar we will need to define some "rules".

In addition to just keeping all these containers in seaparate projects like Scylla and Redpanda, I can think of three options of different levels of integration into Seastar:

1. We could allow in Seastar chunked containers that mimic existing C++ STL containers, with the same APIs and complexity guarantees (with the above caveats...). E.g., a chunked_vector, chunked_map or chunked_unordered_map. 
2. Allow in a Seastar **subdirectory** of container implementations, which is only half-way "in Seastar". Perhaps it will have a less API stability guarantees.
3. Create a separate Github project, of "chunked containers", independent from Seastar and not requiring anything from Seastar. Seastar users can also pick up that library.

**Comment by niekbouman** [comment]
> In general I don't want Seastar to become a collections library.

> 3. Create a separate Github project, of "chunked containers", independent from Seastar and not requiring anything from Seastar. Seastar users can also pick up that library.

I would think that a separate GitHub project could make sense, with even a broader scope (not just containers) where third parties (like RedPanda and others) can contribute implementations under a liberal open-source license, suitable for use with Seastar, and benefit from each other. Maybe this Github project is merely a curated list of links to repos that contain the actual code and that are owned/managed by the contributors  (similar to https://github.com/dotnwat/awesome-seastar)

For example, we have made some adaptations to the Cap'n Proto library, so that we can read/write Cap'n Proto messages from and to a `seastar::input_stream` resp. `seastar::output_stream`. We'd be happy to contribute this, and are currently also writing a `chunked_vector`-like implementation (for which the same holds).

**Comment by travisdowns** [comment]
Thanks @nyh for all your thoughts. I agree with many of the sentiments.

One's one key aspect though that I think helps answer some of the questions or at least means we need a combination approaches is: whether the use of the container is by the application using seastar or is within an unmodifiable part of seastar itself.

For the second case (containers that are used by seastar) I would favor approach (1) or (2), since those are sort of integral parts of seastar. So would you be happy with approach (2) for that case?

Currently we do have crashes in test and production because of the specific metrics aggregation issue referenced above. We can't fix this in Redpanda, short of reducing the number of metrics in the family (and in this case are is a dimensions we don't fully control, since some of the labels depend on the way the end user configures and uses the system). So we are very motivated to contribute to at least a partial solution here.

**Comment by travisdowns** [comment]
xref Avi opening the door ever so slightly to seastar becoming a collections library: https://github.com/scylladb/seastar/pull/2647#issuecomment-2666481524

**Comment by WillemKauf** [comment]
I see a lot of oversized allocations in `redpanda` stemming from the use of `std::unordered_map` in `prometheus-impl.hh`.

https://github.com/scylladb/seastar/blob/510f3148700a88456948a235431978f62eaaad57/src/core/prometheus-impl.hh#L106-L119

Does ScyllaDB not hit this? I can't imagine `redpanda` has orders of magnitude more metrics to scrape than Scylla.

Either way, I think it is another good case for adding a `chunked_map` in `seastar`, whatever that looks like.

**Comment by travisdowns** [comment]
> I can't imagine redpanda has orders of magnitude more metrics to scrape than Scylla.

I can imagine it :). Or they don't use aggregation as we do (which is the source of this particular large alloc IIRC).

In any case, I think @avikivity was open to actually adding something like a chunked map, sort a reconsidering of the position since otherwise (a) everyone effectively has to "BYO" chunked map to use sesatar for anything non-trivial and (b) in any case we need to fix large map cases which arise internally in seastar, so even BYO doesn't work there.

**Comment by StephanDollberg** [comment]
@WillemKauf 

If you see oversized allocs in that aggregation map that usually means we aggregate on labels that don't aggregate away anything (i.e.: bug).

**Comment by avikivity** [comment]
Yes. Seastar practically imposes chunking on its users, and chunked containers aren't common in the wild, so a lot of duplication of effort happens.

Looking at our chunked_vector, it has some non-standard interfaces:

 - reserve_partial() - related to chunking so can be justified
 - memory_size(), external_memory_usage() - scylladb specific


We need to set bounds for container interfaces. Matching the standard library is too restricting, but opening it wide for anything is too relaxed.

**Comment by mykaul** [comment]
Scylla also uses `set_skip_when_empty()` a lot, whereas RedPanda doesn't seem to. May or may not help.

**Comment by travisdowns** [comment]
> Scylla also uses `set_skip_when_empty()` a lot, whereas RedPanda doesn't seem to. May or may not help.

I did look at this, but in our case I don't think so.

**Comment by nyh** [comment]
I think we need to continue this discussion and do something about it.

Today we noticed in the ScyllaDB project, and not for the first time, the need for a hash table that doesn't do very large contiguous allocations for its arrays. See https://github.com/scylladb/scylladb/pull/30383 for the context.

Because avoiding contiguous allocations is not an issue that anyone cares about besides Seastar users, none of the container libraries we already use - like STL, Boost, or Abseil -  have such container with such a feature.

So I think we definitely need to start adding "chunked" versions of containers (vector, unordered_map, etc.) to Seastar. No other container library will ever do this for us. I think it should be an integral part of Seastar, and not some additional library, because basically every Seastar user will need it (even if they don't know yet).

As I (and also Avi) mentioned above, the question is what interfaces the new containers will have, and how much they will be allowed to differ from the STL ones. Another question is how strict the allocation-size guarantee needs to be. Do the containers need to guarantee 128KB allocations are enough for any size - even a terabyte - or for efficiency we can guarantee 128KB allocations only up to  a certain size, e.g., IIRC ScyllaDB's chunked_vector only makes the 128KB contiguous allocation guarantee if the vector has less than 2GB. Maybe we should just adopt a pragmatic solution: The first developer that wants to offer a chunked container will need to convince the other reviewers that their specific decisions make sense, not just for today's ad-hoc needs but also in the long run. Because we don't want to have 7 different versions of each container in Seastar, so whatever implementation we decide to accept will probably be with us for years.

**Comment by avikivity** [comment]
It's okay to have a container that can span hundredss of megabytes (say with a single level of indirection), and if someone want to support tens of gigabytes they can add that later by generalizing the indirection to a tree.

**Comment by avikivity** [comment]
Note that a chunked hash map is very challenging. Not because hash maps are themselves challenging, but because there have been many advances in this area and it will be sad not to adopt them. So the implementer will have to conduct a review of the state of the art before deciding on an approach.

Some of the approaches give up some STL features like pointer stability, also something we'll need to consider.
